### PR TITLE
v6 Schema; move label value out of range key, and re-enable chunk timestamp optimisation.

### LINKED ISSUE #316: ValidationErrors while shutting down ingester
When shutting down an ingester for routine maintenance, I noticed an increase in chunk store flush failures due to "ValidationException":

screen shot 2017-03-07 at 12 10 55

Also, the series queue flush length stalled, rather than steadily decreased

screen shot 2017-03-07 at 12 11 36

Checking the logs, I saw lots of errors like these:

```
request id: DSCAO85A35JL2HLMUCPRPOMTSJVV4KQNSO5AEMVJF66Q9ASUAAJG" source="ingester.go:516"
time="2017-03-07T12:06:28Z" level=error msg="Failed to flush user: ValidationException: One or more parameter values were invalid: Aggregated size of all range keys has exceeded the size limit of 1024 bytes\n\tstatus code: 400, request id: O3EBMNFFUCSU252KDI1TKHTE9NVV4KQNSO5AEMVJF66Q9ASUAAJG"
```

This represents a serious problem, as we cannot shut down these ingesters cleanly.

### PR DESCRIPTION
**Pull request by tomwilkie**
Fixes #316 (ValidationError because range value was too long) and reenabled the v5 query optimisation with the fixed timestamps.

### CODE REVIEW

**Comment by tomwilkie** [comment]
Question: should we impose a limit of label name and value length?

**Comment by juliusv** [comment]
@tomwilkie Is the linked #199 correct? That's already fixed. I guess you meant https://github.com/weaveworks/cortex/issues/316.

**Comment by tomwilkie** [comment]
Yeah I meant #316

**Comment by juliusv** [comment]
> Question: should we impose a limit of label name and value length?

Prometheus has no limit, but it also has a "you trash your server, your problem" mentality. We should probably have limits. Maybe 1k for label names and 4k for label values? Those are already sizes nobody should reach, but who knows...

File: chunk/schema_test.go [hunk 1]
```diff
@@ -392,7 +392,8 @@ func TestSchemaRangeKey(t *testing.T) {
 		base64Keys    = v3Schema(cfg)
 		labelBuckets  = v4Schema(cfg)
 		tsRangeKeys   = v5Schema(cfg)
-		metric        = model.Metric{
+		//v6RangeKeys   = v6Schema(cfg)
```
Comment (juliusv) [inline chunk/schema_test.go]: Should this and the TODO below be cleaned up in this PR?

Comment (tomwilkie) [inline chunk/schema_test.go]: Ah no it was a reminder to me to write some tests.

File: chunk/schema.go [hunk 2]
```diff
@@ -51,6 +53,9 @@ type IndexEntry struct {
 	// For writes, RangeValue will always be set.
 	RangeValue []byte
 
+	// New for v6 schema, Value is not written as part of the range key.
```
Comment (juliusv) [inline chunk/schema.go]: Maybe mention that it's the label value (just "Value" is a bit ambiguous)?

File: chunk/schema.go [hunk 3]
```diff
@@ -602,50 +682,71 @@ func decodeBase64Value(bs []byte) (model.LabelValue, error) {
...
 
 	// v5 schema version 4 range key is chunk end time, label value, chunk ID, version
 	case bytes.Equal(components[3], rangeKeyV4):
-		value, err := decodeBase64Value(components[1])
-		return value, string(components[2]), err
+		chunk := Chunk{ID: string(components[2])}
+		labelValue, err := decodeBase64Value(components[1])
+		return chunk, labelValue, err
+
+	// v6 schema added version 5 range keys, which have the label value written in
```
Comment (juliusv) [inline chunk/schema.go]: Clarify that there's still 4 components, but the second one is empty? Like in the comment for v3.

Comment (tomwilkie) [inline chunk/schema.go]: Done.

**Comment by tomwilkie** [comment]
@juliusv PTAL

File: chunk/schema.go [hunk 4]
```diff
@@ -573,6 +588,71 @@ func (v5Entries) GetReadMetricLabelValueEntries(_, _ uint32, tableName, hashKey
...
+}
+
+func (v6Entries) GetReadMetricLabelValueEntries(_, through uint32, tableName, hashKey string, labelName model.LabelName, labelValue model.LabelValue) ([]IndexEntry, error) {
+	encodedThroughBytes := encodeTime(through)
+	return []IndexEntry{
+		{
+			TableName:  tableName,
+			HashValue:  hashKey + ":" + string(labelName),
+			RangeValue: buildRangeKey(encodedThroughBytes),
+			Value:      []byte(labelValue),
```
Comment (juliusv) [inline chunk/schema.go]: Is it necessary to set the `Value` field for reads here? As far as I could find out, it's only ever read for writes (in https://sourcegraph.com/github.com/weaveworks/cortex@2278819b93fc3f63dd48b677f5325b0c514f8742/-/blob/chunk/schema.go#L57:2-57:7).

**Comment by juliusv** [comment]
One last comment, otherwise :+1:

File: chunk/schema.go (hunk 4 above)
Comment (tomwilkie) [inline chunk/schema.go]: No, you're right its not.  I had thought we might do some dynamodb-side filtering to improve performance, but I'll put that on the perf improvment list instead.

**Pull request merged**